File: ce303-advanced-programming/simple-stock-trading-platform/Java/Server/Market.java
import java.util.*;

public class Market {
    private final Map<Integer, Trader> traders = new TreeMap<>();

    public int nextTraderId = 1;
    public boolean assignedStockholder = false;

    public void createTrader(int traderId, int stockholder)
    {
        Trader trader = new Trader(traderId);
        trader.setStockholder(stockholder);
        traders.put(traderId, trader);
    }

    public List<Integer> getListOfTraders() {
        List<Integer> result = new ArrayList<Integer>();

        for (Trader trader : traders.values())
            result.add(trader.getTraderId());

        return result;
    }

    public int checkStockholder(int traderId) throws Exception {
        if (!traders.containsKey(traderId))
            throw new Exception("Trader " + traderId + " + is not a valid trader, or the client has disconnected.");

        return traders.get(traderId).isStockholder();
    }

    public void transferStock(int traderId, int traderIdFrom, int traderIdTo) throws Exception {
        synchronized (traders)
        {
            if (!traders.containsKey(traderId))
                throw new Exception("Trader " + traderId + " is not a valid trader, or the client has disconnected.");
            if (!traders.containsKey(traderIdFrom))
                throw new Exception("Trader " + traderIdFrom + " is not a valid trader, or the client has disconnected.");
            if (!traders.containsKey(traderIdTo))
                throw new Exception("Trader " + traderIdTo + " is not a valid trader, or the client has disconnected.");
            if (traders.get(traderIdFrom).isStockholder() == 0)
                throw new Exception("Trader " + traderId + " is not the current stockholder.");
            if (traderId != traderIdFrom)
                throw new Exception("Trader can only transfer their stock.");
            traders.get(traderIdFrom).setStockholder(0);
            traders.get(traderIdTo).setStockholder(1);
            System.out.println("Stock transferred to trader " + traderIdTo + ".");
        }
    }

    public void removeTrader(int traderId) {
        if (traders.get(traderId).isStockholder() == 1) {
            if (traders.size() >= 2) {
                traders.get(traderId).setStockholder(0);
                for (Trader trader : traders.values())
                    if (traderId != trader.getTraderId()) {
                        traders.get(trader.getTraderId()).setStockholder(1);
                        System.out.println("Stockholder trader " + traderId + " disconnected. " +
                                "Stock transferred to trader " + trader.getTraderId() + ".");
                        break;
                    }
            } else {
                assignedStockholder = false;
                System.out.println("Stockholder trader " + traderId + " disconnected. " +
                        "No other traders connected: Stock returned to market.");
            }
        }
        traders.remove(traderId);
    }
}

File: ce303-advanced-programming/simple-stock-trading-platform/C#/Server/Market.cs
using System;
using System.Collections.Generic;

namespace Server
{
    public class Market
    {
        private readonly Dictionary<int, Trader> traders = new Dictionary<int, Trader>();

        public int nextTraderId = 1;
        public bool assignedStockholder = false;

        public void createTrader(int traderId, int stockholder)
        {
            Trader trader = new Trader(traderId);
            trader.setStockholder(stockholder);
            traders.Add(traderId, trader);
        }

        public int[] getListOfTraders() {
            List<int> result = new List<int>();

            foreach (Trader trader in traders.Values)
                result.Add(trader.getTraderId());

            return result.ToArray();
        }

        public int checkStockholder(int traderId) {
            if (!traders.ContainsKey(traderId))
                throw new Exception("Trader " + traderId + " + is not a valid trader, or the client has disconnected.");

            return traders[traderId].isStockholder();
        }

        public void transferStock(int traderId, int traderIdFrom, int traderIdTo) {
            lock (traders)
            {
                if (!traders.ContainsKey(traderId))
                    throw new Exception("Trader " + traderId + " is not a valid trader, or the client has disconnected.");
                if (!traders.ContainsKey(traderIdFrom))
                    throw new Exception("Trader " + traderIdFrom + " is not a valid trader, or the client has disconnected.");
                if (!traders.ContainsKey(traderIdTo))
                    throw new Exception("Trader " + traderIdTo + " is not a valid trader, or the client has disconnected.");
                if (traders[traderIdFrom].isStockholder() == 0)
                    throw new Exception("Trader " + traderId + " is not the current stockholder.");
                if (traderId != traderIdFrom)
                    throw new Exception("Trader can only transfer their stock.");
                traders[traderIdFrom].setStockholder(0);
                traders[traderIdTo].setStockholder(1);
                Console.WriteLine("Stock transferred to trader " + traderIdTo + ".");
            }
        }

        public void removeTrader(int traderId) {
            if (traders[traderId].isStockholder() == 1) {
                if (traders.Count >= 2) {
                    traders[traderId].setStockholder(0);
                    foreach (Trader trader in traders.Values)
                        if (traderId != trader.getTraderId()) {
                            traders[trader.getTraderId()].setStockholder(1);
                            Console.WriteLine("Stockholder trader " + traderId + " disconnected. " +
                                    "Stock transferred to trader " + trader.getTraderId() + ".");
                            break;
                        }
                } else {
                    assignedStockholder = false;
                    Console.WriteLine("Stockholder trader " + traderId + " disconnected. " +
                            "No other traders connected: Stock returned to market.");
                }
            }
            traders.Remove(traderId);
        }
    }
}